library/core/src/iter/traits/iterator.rs

```diff
@@ -1670,6 +1672,77 @@ pub trait Iterator {
         FromIterator::from_iter(self)
     }
 
+    /// Collects all items from the iterator into an array of a specific size.
+    ///
+    /// If the number of elements inside the iterator is exactly equal to the
+    /// array size, then the array is returned inside `Some`, otherwise `None`
+    /// is returned.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(iter_collect_array)]
+    ///
+    /// let xs = [1, 2, 3];
+    /// let iter = xs.iter();
+    /// let [_a, b, _c] == iter.clone().collect_array().unwrap();
+    /// assert_eq!(b, 2);
+    ///
+    /// match iter.collect_array() {
+    ///     Some([_, _]) => panic!("Didn't expect to see two only elements"),
+    ///     None => (),
+    /// }
+    /// ```
+    #[inline]
+    #[unstable(reason = "new API", issue = "none", feature = "iter_collect_array")]
+    fn collect_array<const N: usize>(self) -> Option<[Self::Item; N]>
+    where
+        Self: Sized,
+    {
+        struct Buf<T, const N: usize> {
```

Comment:
Do we have this type somewhere already?

Reply:
I guess there is one in `[T; N]::map()` method

Reply:
Looks like there are more copies in other PRs, like https://github.com/rust-lang/rust/pull/79451.

Maybe it'd be worth waiting for https://github.com/rust-lang/rust/pull/75644?  Then this might essentially just be `[T; N]::try_generate(|_| it.next())`...

Reply:
Seems like we need `core::collections::array_vec` already :D

Reply:
There's an item for the next libs meeting to talk about that https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/PRs.20need.20a.20decision.20on.20.60try_generate.60.20or.20.60ArrayVec.60.20or.20.2E.2E.2E/near/218779392 :)

Reply:
I'm currently writing  `Iterator::array_chunks` and I have +- the same code :) 

(though it's a bit more convoluted since `DoubleEndedItetator` impl requires an ability to push_back)

Reply:
Unfortunately we rarely get to our _needs decision_ items 🙃 

If we keep tabs on these unwind-safe-partially-initialized-array types we're introducing I'd be happy not to block any of these PRs individually on them and run a pass over later to replace them with something more general.